test/e2e/clusternetworkpolicy_test.go

```diff
@@ -0,0 +1,558 @@
+		nil, nil, secv1alpha1.RuleActionDrop)
+
+	builder2 := &ClusterNetworkPolicySpecBuilder{}
+	// cnp-allow will also apply to all pods in namespace x
+	builder2 = builder2.SetName("cnp-allow").
+		SetPriority(5).
+		SetAppliedToGroup(nil, map[string]string{"ns": "x"}, nil, nil)
+	builder2.AddEgress(v1.ProtocolTCP, &p80, nil, nil, nil, map[string]string{"ns": "z"},
+		nil, nil, secv1alpha1.RuleActionAllow)
+	// This rule should take no effect as it will be overridden by the first rule of cnp-drop
+	builder2.AddEgress(v1.ProtocolTCP, &p80, nil, nil, nil, map[string]string{"ns": "y"},
+		nil, nil, secv1alpha1.RuleActionAllow)
+
+	// Only egress from pods in namespace x to namespace y should be denied
+	reachabilityBothCNP := NewReachability(allPods, true)
+	reachabilityBothCNP.Expect(Pod("x/a"), Pod("y/a"), false)
+	reachabilityBothCNP.Expect(Pod("x/b"), Pod("y/a"), false)
+	reachabilityBothCNP.Expect(Pod("x/c"), Pod("y/a"), false)
+	reachabilityBothCNP.Expect(Pod("x/a"), Pod("y/b"), false)
+	reachabilityBothCNP.Expect(Pod("x/b"), Pod("y/b"), false)
+	reachabilityBothCNP.Expect(Pod("x/c"), Pod("y/b"), false)
+	reachabilityBothCNP.Expect(Pod("x/a"), Pod("y/c"), false)
+	reachabilityBothCNP.Expect(Pod("x/b"), Pod("y/c"), false)
+	reachabilityBothCNP.Expect(Pod("x/c"), Pod("y/c"), false)
+
+	testStep := []*TestStep{
+		{
+			"Both CNP",
+			reachabilityBothCNP,
+			[]*secv1alpha1.ClusterNetworkPolicy{builder2.Get(), builder1.Get()},
+			80,
+			0,
+		},
+	}
+	testCase := []*TestCase{
+		{"CNP Rule Priority", testStep},
+	}
+	executeTests(t, testCase)
+}
+
+// executeTests runs all the tests in testList and print results
+func executeTests(t *testing.T, testList []*TestCase) {
+	for _, testCase := range testList {
+		log.Infof("running test case %s", testCase.Name)
+		log.Debugf("cleaning-up previous policies and sleeping for %v", networkPolicyDelay)
+		err := k8sUtils.CleanCNPs()
+		time.Sleep(networkPolicyDelay)
+		failOnError(err, t)
+		for _, step := range testCase.Steps {
+			log.Infof("running step %s of test case %s", step.Name, testCase.Name)
+			reachability := step.Reachability
+			for _, cnp := range step.CNPs {
+				if cnp != nil {
+					log.Debugf("creating CNP %v", cnp.Name)
+					_, err := k8sUtils.CreateOrUpdateCNP(cnp)
+					failOnError(err, t)
+				}
+			}
+			if len(step.CNPs) > 0 {
+				log.Debugf("Sleeping for %v for all CNPs to take effect", networkPolicyDelay)
+				time.Sleep(networkPolicyDelay)
+			}
+			start := time.Now()
+			k8sUtils.Validate(allPods, reachability, step.Port)
+			step.Duration = time.Now().Sub(start)
+			reachability.PrintSummary(true, true, true)
+
+			_, wrong, _ := step.Reachability.Summary()
+			if wrong != 0 {
+				t.Errorf("failure -- %d wrong results", wrong)
+			}
+		}
+	}
+	allTestList = append(allTestList, testList...)
+}
+
+// printResults summarizes test results for all the testcases
+func printResults() {
+	fmt.Printf("\n\n---------------- Test Results ------------------\n\n")
```

Comment:
a note for later: we should probably log the summaries and matrices to a file to avoid clogging the logs? At the same time it is convenient to have it in the logs, especially for CI. Something to think about.

Reply:
Agreed. Will take this into consideration for future work